### PR TITLE
random generator update for Mac proposal

### PR DESCRIPTION
**Pull request by devnexen**

### CODE REVIEW

File: aclocal.m4 [hunk 1]
```diff
@@ -43,6 +43,5 @@ m4_include([tool/m4/ruby_stack_grow_direction.m4])
 m4_include([tool/m4/ruby_try_cflags.m4])
 m4_include([tool/m4/ruby_try_cxxflags.m4])
 m4_include([tool/m4/ruby_try_ldflags.m4])
-m4_include([tool/m4/ruby_type_attribute.m4])
```
Comment (shyouhei) [inline aclocal.m4:46]: Tell us the reason for this removal?

File: random.c [hunk 2]
```diff
@@ -421,7 +421,26 @@ random_init(int argc, VALUE *argv, VALUE obj)
 # define USE_DEV_URANDOM 0
 #endif
 
-#if USE_DEV_URANDOM
+#if HAVE_GETENTROPY
+# if defined(__ANDROID__) || defined(__APPLE__)
+#  include <sys/random.h>
+# endif
+# define MAX_SEED_LEN_PER_READ 256
```
Comment (shyouhei) [inline random.c:428]: Why is the reason behind this 256?

**Review by shyouhei — COMMENTED** [review]
Please provide a bit more detailed descriptions.

File: aclocal.m4 (hunk 1 above)
Comment (devnexen) [reply aclocal.m4]: Unexpected change, I ll roll it back.

File: random.c (hunk 2 above)
Comment (devnexen) [reply random.c:428]: It s to avoid `EIO` errno case since getentropy can handle 256 bytes maximum per call.

File: random.c [hunk 3]
```diff
@@ -421,7 +421,26 @@ random_init(int argc, VALUE *argv, VALUE obj)
 # define USE_DEV_URANDOM 0
 #endif
 
-#if USE_DEV_URANDOM
+#if HAVE_GETENTROPY
+# if defined(__ANDROID__) || defined(__APPLE__)
+#  include <sys/random.h>
+# endif
```
Comment (nobu) [inline random.c:427]: There is the code for `<sys/random.h>` already, this should be there too.
That is, at line 52,
```C
#if defined HAVE_GETRANDOM || defined HAVE_GETENTROPY
```

File: random.c (hunk 2 above)
Comment (nobu) [reply random.c:428]: The length is described in the man-pages, but no symbol seems defined for it.

macOS:
```
The maximum buffer size permitted is 256 bytes.  If buflen exceeds this, an error of EIO will be indicated.
```
Linux:
```
The maximum permitted value for the length argument is 256.
```

File: random.c [hunk 4]
```diff
@@ -497,7 +516,8 @@ fill_random_bytes_syscall(void *buf, size_t size, int unused)
 {
 #if (defined(__OpenBSD__) && OpenBSD >= 201411) || \
     (defined(__NetBSD__)  && __NetBSD_Version__ >= 700000000) || \
-    (defined(__FreeBSD__) && __FreeBSD_version >= 1200079)
+    (defined(__FreeBSD__) && __FreeBSD_version >= 1200079) || \
+    defined(__APPLE__)
```
Comment (nobu) [inline random.c:520]: In what cases is this necessary?

Comment (devnexen) [reply random.c:520]: Maybe not ... too much randomness at a time in this case indeed :)

File: random.c (hunk 3 above)
Comment (devnexen) [reply random.c:427]: I just checked if I remove this line it fails to compile

File: random.c (hunk 4 above)
Comment (nobu) [reply random.c]: If this condition is enabled,  this `fill_random_bytes_syscall()` never fails and `fill_random_bytes_urandom()` using `getentropy()` is never used.

Comment (devnexen) [reply random.c]: true I removed it in the last commit anyway.

File: random.c (hunk 3 above)
Comment (nobu) [reply random.c:427]: It means to add ` || defined HAVE_GETENTROPY` to line 52.

Comment (devnexen) [reply random.c:427]: Ah gotcha sorry.

**Comment by mame** [comment]
I guess this change broke OpenBSD build: http://rubyci.s3.amazonaws.com/openbsd-current/ruby-master/log/20210118T023008Z.fail.html.gz

**Comment by mame** [comment]
https://github.com/ruby/ruby/commit/6abf393e8f372464322dc0a8c2bda80f8b389acd

**Comment by rhenium** [comment]
This patch seems to break the fallback to `/dev/urandom` on Linux < 3.17 with glibc >= 2.25: https://bugs.ruby-lang.org/issues/21448

On macOS, it seems we first try `CCRandomGenerateBytes()` or `SecRandomCopyBytes()`, depending on the OS version, and if it fails, we try `getentropy()`. What are the differences among them, and can they ever fail?

Unfortunately, Apple doesn't seem to have the man pages available on the web (anymore?)

**Pull request merged**